var/spack/repos/builtin/packages/sympack/package.py

```diff
@@ -0,0 +1,91 @@
+
+from os import environ as env
+
+from spack.package import *
+
+
+class Sympack(CMakePackage, CudaPackage):
+    """symPACK is a sparse symmetric matrix direct linear solver,
+    with optional support for CUDA devices"""
+
+    homepage = "https://go.lbl.gov/sympack"
+    url = "https://github.com/symPACK/symPACK/archive/refs/tags/v3.0.tar.gz"
+    git = "https://github.com/symPACK/symPACK.git"
+
+    maintainers("bonachea")
+
+    version("master", branch="master")
+
+    version("3.0", sha256="bd04284bb6a309a71cd4f9f54f72345ff98fa7ba5719498df1e8a00ef05c41de")
+    version("2.0.1", sha256="21c172e902531c94c3bb5932c15de4b4ec9adf9c0d8e2071bb12cdbdfa25ca52")
+    version("2.0", sha256="93fcfbadab73718e249e7ed12641a3c6be58b19cafdf6bad12a6a09c3d1eb4a1")
+
+    depends_on("cmake@3.11:", type="build")
+
+    depends_on("upcxx@2019.9.0:")
+    depends_on("upcxx@2022.3.0:+cuda", when="+cuda")
+
+    depends_on("mpi")
+    depends_on("blas")
+
+    variant(
+        "cuda",
+        default=False,
+        when="@3.0:",
+        description="Enables solver to offload large operations to CUDA GPUs",
+    )
+    conflicts("+cuda", when="@:2", msg="symPACK version 3.0 or later required for CUDA support")
+    depends_on("cuda@6.0:", when="+cuda")
+
+    variant("scotch", default=False, description="Enable SCOTCH ordering")
+    depends_on("scotch", when="+scotch")
+
+    variant("ptscotch", default=False, description="Enable PT-SCOTCH ordering")
+    depends_on("scotch+mpi", when="+ptscotch")
+
+    variant("metis", default=False, description="Enable MeTiS ordering")
+    depends_on("metis", when="+metis")
+
+    variant("parmetis", default=False, description="Enable ParMETIS ordering")
+    depends_on("parmetis", when="+parmetis")
+
+    variant(
+        "network",
+        default="auto",
+        multi=False,
+        values=("auto", "smp", "mpi", "ibv", "udp", "ofi", "ucx"),
+        description="The hardware-dependent UPC++ network backend to use. "
+        + "'auto' selects the default network provided by the UPC++ install.",
+    )
+
+    build_targets = ["all", "run_sympack2D"]
+
+    def cmake_args(self):
+        # UPC++ is picky about what C++ compiler we use:
+        meta = join_path(self.spec["upcxx"].prefix.bin, "upcxx-meta")
+        cxx = Executable(meta)("CXX", output=str, error=str).strip()
+        args = [
+            self.define("CMAKE_CXX_COMPILER", cxx),
+            self.define_from_variant("ENABLE_CUDA", "cuda"),
+            self.define_from_variant("ENABLE_SCOTCH", "scotch"),
+            self.define_from_variant("ENABLE_PTSCOTCH", "ptscotch"),
+            self.define_from_variant("ENABLE_METIS", "metis"),
+            self.define_from_variant("ENABLE_PARMETIS", "parmetis"),
+        ]
+        return args
+
+    @run_before("cmake")
+    def set_cmake_env(self):
+        spec = self.spec
+        if "network=auto" not in spec:
+            env["UPCXX_NETWORK"] = spec.variants["network"].value
```

Comment:
Can we use the more idiomatic `setup_build_environment` callback instead of this method?

Reply:
@alalazo FWIW `run_before("cmake")` appears in 36 packages which also seems pretty idiomatic. Perhaps the [CMake packaging guide](https://spack.readthedocs.io/en/latest/build_systems/cmakepackage.html) should explain this is the preferred mechanism for setting the environment for the `cmake` step? (which notably occurs *before* CMakePackage's "build" step)

Requested change applied in 041aa27ba2, and I've verified it works as expected.

Reply:
It's more idiomatic for setting environment variables, documentation is [here](https://spack.readthedocs.io/en/latest/packaging_guide.html#runtime-and-build-time-environment-variables). If you had to do anything else before the "cmake" phase, then using the decorator would have been absolutely fine.

Reply:
@alalazo Thanks for the clarification. I was familiar with the documentation you linked, but I wasn't certain it would be effective for the "cmake" stage (which precedes the cmake "build" stage) until I tested it at your urging.

In any case, I believe this PR is ready to merge.